### PR TITLE
inline binary comparison ops to prevent function call overhead

### PR DESCRIPTION
**Pull request by richardstartin**
## Description

In a CPU bound load test where there are at least 3 binary operations in play, nearly 10% of the time on the server is spent evaluating `BinaryOperatorTransformFunction.getIntResult`. 

Screenshot 2021-11-05 at 13 08 48

This moves the very simple overrides into the base class, and leaves the individual classes in place for the sake of the registration mechanism alone. 

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java [hunk 1]
```diff
@@ -71,218 +113,331 @@ public TransformResultMetadata getResultMetadata() {
...
+    for (int i = 0; i < length; i++) {
+      _results[i] = getIntResult(leftStringValues[i].compareTo(rightStringValues[i]));
+    }
+  }
+
+  private void fillResultBytes(ProjectionBlock projectionBlock, int length) {
+    byte[][] leftBytesValues = _leftTransformFunction.transformToBytesValuesSV(projectionBlock);
+    byte[][] rightBytesValues = _rightTransformFunction.transformToBytesValuesSV(projectionBlock);
+    for (int i = 0; i < length; i++) {
+      _results[i] = getIntResult((new ByteArray(leftBytesValues[i])).compareTo(new ByteArray(rightBytesValues[i])));
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java:256]: Avoid creating the object
```suggestion
      _results[i] = getIntResult(ByteArray.compare(leftBytesValues[i], rightBytesValues[i]));
```

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java [hunk 2]
```diff
@@ -71,218 +113,331 @@ public TransformResultMetadata getResultMetadata() {
...
+      case GREATER_THAN:
+        return comparisonResult > 0;
+      case LESS_THAN:
+        return comparisonResult < 0;
+      case LESS_THAN_OR_EQUAL:
+        return comparisonResult <= 0;
+      case NOT_EQUAL:
+        return comparisonResult != 0;
+      default:
+        assert false;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java:439]: Throw exception?

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java [hunk 3]
```diff
@@ -71,218 +113,331 @@ public TransformResultMetadata getResultMetadata() {
...
+    long[] rightValues = _rightTransformFunction.transformToLongValuesSV(projectionBlock);
+    for (int i = 0; i < length; i++) {
+      _results[i] = getIntResult(Long.compare(leftValues[i], rightValues[i]));
+    }
+  }
+
+  private void fillFloatResultArray(ProjectionBlock projectionBlock, long[] leftValues, int length) {
+    float[] rightFloatValues = _rightTransformFunction.transformToFloatValuesSV(projectionBlock);
+    for (int i = 0; i < length; i++) {
+      _results[i] = getIntResult(Double.compare(leftValues[i], rightFloatValues[i]));
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java:317]: We should use `BigDecimal` for this comparison

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java [hunk 4]
```diff
@@ -71,218 +113,331 @@ public TransformResultMetadata getResultMetadata() {
...
+    }
+  }
+
+  private void fillDoubleResultArray(ProjectionBlock projectionBlock, long[] leftValues, int length) {
+    double[] rightDoubleValues = _rightTransformFunction.transformToDoubleValuesSV(projectionBlock);
+    for (int i = 0; i < length; i++) {
+      _results[i] = getIntResult(Double.compare(leftValues[i], rightDoubleValues[i]));
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java:324]: Same here

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java (hunk 2 above)
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java:439]: This is impossible unless this class is edited, I don't see the point in handling it any other way. The tests will fail with an assertion error anyway.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java (hunk 1 above)
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java:256]: This was lifted and shifted, but it makes sense to change it anyway.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java (hunk 3 above)
Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java:317]: I don't think this is necessary. `BigDecimal` is ridiculously expensive. Do you have a particular case in mind?

Comment (richardstartin) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/BinaryOperatorTransformFunction.java:317]: I assume you are concerned about inexact translation to `double` - we now avoid converting to `BigDecimal` if the absolute value of the `long` is smaller than 2^53.

**Pull request merged**